### PR TITLE
bpfd: Move image manager into dedicated thread

### PR DESCRIPTION
**Pull request by dave-tucker**
This commit refactors the ImageManager to be more self contained.

### CODE REVIEW

File: bpfd/src/oci_utils/image_manager.rs [hunk 1]
```diff
@@ -67,37 +78,118 @@ impl BytecodeImage {
...
+                        },
+                        Command::GetBytecode { path, resp } => {
+                            let result = self.get_bytecode_from_image_store(path).await;
+                            let _ = resp.send(result);
+                        }
+                    }
+                }
+            }
+        }
+        info!("image_manager: Stopping processing commands");
```
Comment (astoycos) [inline bpfd/src/oci_utils/image_manager.rs:162]: NIT
```suggestion
        info!("image_manager: Stopped processing commands");
```

File: bpfd/src/bpf.rs [hunk 2]
```diff
@@ -185,7 +198,10 @@ impl BpfManager {
     ) -> Result<Uuid, BpfdError> {
         debug!("BpfManager::add_multi_attach_program()");
 
-        let program_bytes = program.data_mut().program_bytes().await?;
+        let program_bytes = program
```
Comment (astoycos) [inline bpfd/src/bpf.rs:201]: Nit can you move this into add_program() so that we don't have to call the same code twice in add_multi && add_single etc

File: bpfd/src/oci_utils/image_manager.rs [hunk 3]
```diff
@@ -67,37 +78,118 @@ impl BytecodeImage {
...
+    Pull {
+        image: String,
+        pull_policy: ImagePullPolicy,
+        username: Option<String>,
+        password: Option<String>,
+        resp: Responder<Result<(String, String), ImageError>>,
+    },
+    GetBytecode {
+        path: String,
+        resp: Responder<Result<Vec<u8>, anyhow::Error>>,
```
[the remark this answers is not in the document]
Comment (astoycos) [reply bpfd/src/oci_utils/image_manager.rs:121]: Also after reading more... Should we just return the PATH to the bytecode here on disk rather than the raw bytes? Then in the core BpfManager we can always just use `crate::utils::read(l)` on the file.  It'd also mean less data going over the channel

File: bpfd/src/multiprog/xdp.rs [hunk 4]
```diff
@@ -75,11 +77,32 @@ impl XdpDispatcher {
             None,
             None,
         );
-        let (path, section_name) = image
-            .get_image(None)
+        let (tx, rx) = oneshot::channel();
```
Comment (astoycos) [inline bpfd/src/multiprog/xdp.rs:80]: I think we should consider pre-pulling the dispatcher images at bpfd startup rather than attempting to every time Dispatcher::new is called

but we can do that in a follow up.

**Review by astoycos — COMMENTED** [review]
Pretty much there just some nits/questions I like this design!!a

File: bpfd/src/oci_utils/image_manager.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dave-tucker) [reply bpfd/src/oci_utils/image_manager.rs:121]: When I started this work I wanted to collapse the two API commands Pull and GetBytecode into one.
We can’t call `crate::utils::read` the path that comes back from `Pull` since its a directory in our image store.
The `GetBytecode` API does all the fancy things from that parh - fetch manifest, find layers, ungzip them and return the bytes etc…

ISTR we discussed having a cache of inflated layers on disk, but decided against it because without some form of persistent database we wouldn’t know if any of those layers had been tampered with - i.e a malicious actor replaces the bytecode while bpfd is running and/or restarting.

My understanding of tokio channels is that passing Vec’s is cheap, because a Vec is just a pointer (the data is on the heap). Since we're copying the pointer, not the heap data, this should be fine 🤔 

In other words, this is the least-expensive short-term option 😆

File: bpfd/src/multiprog/xdp.rs [hunk 5]
```diff
@@ -75,11 +77,32 @@ impl XdpDispatcher {
...
+            })
+            .await
+            .map_err(|e| BpfdError::BpfBytecodeError(e.into()))?;
+
+        let (path, section_name) = rx
             .await
+            .map_err(|e| BpfdError::BpfBytecodeError(e.into()))?
             .map_err(|e| BpfdError::BpfBytecodeError(e.into()))?;
-        let program_bytes = get_bytecode_from_image_store(path).await?;
+
```
[the remark this answers is not in the document]
Comment (dave-tucker) [reply bpfd/src/multiprog/xdp.rs:96]: Not sure I understand the comment. Can you clarify?

[the remark this answers is not in the document]
Comment (astoycos) [reply bpfd/src/multiprog/xdp.rs:96]: Sorry.. I see two identical `map_err` statements 

https://github.com/bpfd-dev/bpfd/blob/e95003724005baf783e96374d6686f256852ac3a/bpfd/src/multiprog/xdp.rs#L94C30-L94C30

It seemed weird

[the remark this answers is not in the document]
Comment (astoycos) [reply bpfd/src/multiprog/xdp.rs:96]: Not going to block the merge on this

File: bpfd/src/oci_utils/image_manager.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (astoycos) [reply bpfd/src/oci_utils/image_manager.rs:121]: > My understanding of tokio channels is that passing Vec’s is cheap, because a Vec is just a pointer (the data is on the heap). Since we're copying the pointer, not the heap data, this should be fine 

I like it sounds good to me

File: bpfd/src/multiprog/xdp.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (dave-tucker) [reply bpfd/src/multiprog/xdp.rs:96]: ah yeah that's intentional, we're dealing with a `Result<Result<(String, String), ImageError>>`
The first is `Recv error` at the channel layer, the latter is from the `ImageManager` itself.

**Review by astoycos — APPROVED** [review]
LGTM

Thanks!
